Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h

```diff
@@ -0,0 +1,62 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <memory>
+#include "Common/CommonTypes.h"
+#include "Core/HW/HSP/HSP_Device.h"
+
+namespace HSP
+{
+class CHSPDevice_GBPlayer;
+
+class IGBPlayer
+{
+public:
+  IGBPlayer(CHSPDevice_GBPlayer*);
+  virtual ~IGBPlayer();
+
+  virtual void Reset() = 0;
+  virtual void Shutdown();
+
+  virtual bool IsLoaded() const = 0;
+  virtual bool IsGBA() const = 0;
+
+  virtual bool LoadGame(const std::string& path) = 0;
+  virtual bool LoadBootrom(const std::string& path) = 0;
+
+  virtual void ReadScanlines(u32*) = 0;
+  virtual void ReadAudio(u8*) = 0;
+
+  virtual void SetKeys(u16) = 0;
+
+protected:
+  CHSPDevice_GBPlayer* const m_player;
+};
+
+class CHSPDevice_GBPlayer : public IHSPDevice
+{
+public:
+  enum class IRQ : int;
+
+  CHSPDevice_GBPlayer(HSPDevices device);
```

Comment:
```suggestion
  explicit CHSPDevice_GBPlayer(HSPDevices device);
```

The destructor should also be defaulted in the cpp file, since the class contains a non-trivial member like a std::unique_ptr. This allows forward-declaring the class, should it be necessary, without unexpected/confusing compilation errors.

Reply:
What do you mean by "defaulted in the cpp file"? Is there some C++11 defaulting behavior I'm unfamiliar with?

Reply:
e.g. doing the following:

```cpp
// header

class Thing {
public:
    ~CHSPDevice_GBPlayer();
};

// cpp file:

CHSPDevice_GBPlayer::~CHSPDevice_GBPlayer() = default;
```